### PR TITLE
[Diagnostics] Add verifier support for educational notes

### PR DESCRIPTION
**Pull request by owenv**
Builds on #28313, so only the last two commits are new. This adds support to the verifier for verifying educational notes are emitted via `{{educational-notes=some-note}}` syntax.

### CODE REVIEW

File: tools/sil-opt/SILOpt.cpp [hunk 1]
```diff
@@ -498,12 +504,12 @@ int main(int argc, char **argv) {
 
   bool HadError = CI.getASTContext().hadError();
 
+  PrintDiags.setSuppressOutput(false);
+
   // If we're in -verify mode, we've buffered up all of the generated
   // diagnostics.  Check now to ensure that they meet our expectations.
```
Comment (brentdax) [inline tools/sil-opt/SILOpt.cpp]: Is this comment still accurate with the refactoring you've done? I don't think the final processing actually happens here.

File: include/swift/Frontend/DiagnosticVerifier.h [hunk 2]
```diff
@@ -18,22 +18,124 @@
...
-  /// files corresponding to the specified \p BufferIDs and tear down our
-  /// support for capturing and verifying diagnostics.
-  ///
-  /// This returns true if there are any mismatches found.
-  bool verifyDiagnostics(SourceManager &SM, ArrayRef<unsigned> BufferIDs,
-                         bool autoApplyFixes, bool ignoreUnknown);
+/// This class implements support for -verify mode in the compiler.  It
+/// buffers up diagnostics produced during compilation, then checks them
+/// against expected-error markers in the source file.
+class DiagnosticVerifier : public DiagnosticConsumer {
```
Comment (brentdax) [inline include/swift/Frontend/DiagnosticVerifier.h]: Do we know why `DiagnosticVerifier` wasn't a `DiagnosticConsumer` before? I'm worried that this change might make the verifier not, for instance, compose with other diagnostic consumers the way it used to. Do `-serialize-diagnostics-path` and `-emit-fixits-path` in combination with `-verify` work the same way they used to? 

Separately: Why is `DiagnosticVerifier` fully defined in an `include/` header? The other diagnostic consumers only declare a function to create an instance and hide the rest of the implementation, and the old `-verify` implementation also only exposed a few functions to set it up. Is there an architectural difference that forces `DiagnosticVerifier` to be different, or was this simply an oversight?

File: tools/sil-opt/SILOpt.cpp (hunk 1 above)
Comment (brentdax) [inline tools/sil-opt/SILOpt.cpp]: Is this comment still accurate with the refactoring you've done? I don't think the final processing actually happens here.

**Review by beccadax — COMMENTED** [review]
Before we get down to brass tacks, I have a couple big-picture questions about the refactoring that underlies this PR.

**Comment by futurejones** [comment]
Can one of the admins verify this patch?

**Comment by owenv** [comment]
Rebasing this on the latest changes from https://github.com/apple/swift/pull/28313. Some of the tests can probably be unified now, but let's make sure everything still passes.

**Comment by owenv** [comment]
@swift-ci please smoke test

**Comment by owenv** [comment]
This has fallen out of date with #28313, I'll open a new PR once that one lands

**Pull request closed without merging**